### PR TITLE
get_dirty_log() fix

### PR DESCRIPTION
**Pull request by serban300**
call sysconf() to get the page size

### CODE REVIEW

File: src/ioctls/vm.rs [hunk 1]
```diff
@@ -791,22 +791,24 @@ impl VmFd {
...
+        };
 
-        let div_round_up = |dividend, divisor| (dividend + divisor - 1) / divisor;
         // For ease of access we are saving the bitmap in a u64 vector. We are using ceil to
-        // make sure we count all dirty pages even when `mem_size` is not a multiple of
-        // page_size * 64.
-        let bitmap_size = div_round_up(memory_size, page_size * 64);
+        // make sure we count all dirty pages even when `memory_size` is not a multiple of
+        // `page_size * 64`.
+        let div_ceil = |dividend, divisor| (dividend + divisor - 1) / divisor;
```
Comment (alxiord) [inline src/ioctls/vm.rs:804]: Why replace `div_round_up`?

File: src/ioctls/vm.rs [hunk 2]
```diff
@@ -791,22 +791,24 @@ impl VmFd {
     ///
     pub fn get_dirty_log(&self, slot: u32, memory_size: usize) -> Result<Vec<u64>> {
         // Compute the length of the bitmap needed for all dirty pages in one memory slot.
-        // One memory page is 4KiB (4096 bits) and `KVM_GET_DIRTY_LOG` returns one dirty bit for
+        // One memory page is `page_size` bits and `KVM_GET_DIRTY_LOG` returns one dirty bit for
         // each page.
-        let page_size = 4 << 10;
+        let page_size = match unsafe { libc::sysconf(libc::_SC_PAGESIZE) } as isize {
```
Comment (andreeaflorescu) [inline src/ioctls/vm.rs]: `as isize` is not needed.

File: src/ioctls/vm.rs [hunk 3]
```diff
@@ -791,22 +791,24 @@ impl VmFd {
...
 
-        let div_round_up = |dividend, divisor| (dividend + divisor - 1) / divisor;
         // For ease of access we are saving the bitmap in a u64 vector. We are using ceil to
-        // make sure we count all dirty pages even when `mem_size` is not a multiple of
-        // page_size * 64.
-        let bitmap_size = div_round_up(memory_size, page_size * 64);
+        // make sure we count all dirty pages even when `memory_size` is not a multiple of
+        // `page_size * 64`.
+        let div_ceil = |dividend, divisor| (dividend + divisor - 1) / divisor;
+        let bitmap_size = div_ceil(memory_size, page_size * 64);
```
Comment (andreeaflorescu) [inline src/ioctls/vm.rs:805]: Should we check that `page_size` is not 0?

**Review by andreeaflorescu — COMMENTED** [review]
Please use a more descriptive commit message to explain what was the problem that you are fixing with your commit. This is important to for properly keeping track of changes.

File: src/ioctls/vm.rs (hunk 1 above)
Comment (serban300) [reply src/ioctls/vm.rs:804]: I just used the same name as in the rust-num crate: https://docs.rs/num/0.2.1/num/integer/fn.div_ceil.html

File: src/ioctls/vm.rs (hunk 2 above)
Comment (serban300) [reply src/ioctls/vm.rs]: Fixed.

File: src/ioctls/vm.rs (hunk 3 above)
Comment (serban300) [reply src/ioctls/vm.rs:805]: I don't think the page_size can be configured to 0.

**Comment by serban300** [comment]
> Please use a more descriptive commit message to explain what was the problem that you are fixing with your commit. This is important to for properly keeping track of changes.

@andreeaflorescu Done. Please take another look.

File: src/ioctls/vm.rs (hunk 3 above)
Comment (alxiord) [reply src/ioctls/vm.rs:805]: [That's right](http://man7.org/linux/man-pages/man3/sysconf.3.html):
```
       PAGESIZE - _SC_PAGESIZE
              Size of a page in bytes.  Must not be less than 1.
```

**Review by alxiord — DISMISSED** [review]

**Review by andreeaflorescu — CHANGES_REQUESTED** [review]
Temporarily blocking the merge of this PR to double check that we calculate the page size correctly.

File: src/ioctls/vm.rs [hunk 4]
```diff
@@ -791,22 +791,24 @@ impl VmFd {
     ///
     pub fn get_dirty_log(&self, slot: u32, memory_size: usize) -> Result<Vec<u64>> {
         // Compute the length of the bitmap needed for all dirty pages in one memory slot.
-        // One memory page is 4KiB (4096 bits) and `KVM_GET_DIRTY_LOG` returns one dirty bit for
+        // One memory page is `page_size` bits and `KVM_GET_DIRTY_LOG` returns one dirty bit for
```
Comment (andreeaflorescu) [inline src/ioctls/vm.rs]: s/bits/bytes/

File: src/ioctls/vm.rs [hunk 5]
```diff
@@ -791,22 +791,24 @@ impl VmFd {
...
+            -1 => return Err(errno::Error::last()),
+            ps => ps as usize,
+        };
 
-        let div_round_up = |dividend, divisor| (dividend + divisor - 1) / divisor;
         // For ease of access we are saving the bitmap in a u64 vector. We are using ceil to
-        // make sure we count all dirty pages even when `mem_size` is not a multiple of
-        // page_size * 64.
-        let bitmap_size = div_round_up(memory_size, page_size * 64);
+        // make sure we count all dirty pages even when `memory_size` is not a multiple of
```
Comment (andreeaflorescu) [inline src/ioctls/vm.rs:802]: This was wrong before as well, but to enforce that the vector has u64 elements, we should define it as:
```vec![0u64; bitmap_size];```

Can you also do this change even though it's not necessarily related to this PR?

File: src/ioctls/vm.rs (hunk 4 above)
Comment (serban300) [reply src/ioctls/vm.rs]: Fixed.

File: src/ioctls/vm.rs (hunk 5 above)
Comment (serban300) [reply src/ioctls/vm.rs:802]: Fixed.

**Review by andreeaflorescu — DISMISSED** [review]

**Review by alxiord — APPROVED** [review]

**Review by andreeaflorescu — APPROVED** [review]

**Pull request merged**